### ISSUE TITLE
Incorrect measurements during parallel requests with profiling examples

### ISSUE THREAD

**Issue by csvn** [middyjs/middy#1047]
**Describe the bug**
The documentation regarding [Profiling](https://middy.js.org/docs/best-practices/profiling/) is misleading. It does not mention that the time and memory cannot be measured correctly if requests come in parallel.

I don't know if the profiling examples are intended only for lab environments or to examples of real code deployed on lambda. If they are real, they will likely produce incorrect measurements if there is any overlap between requests.

Suggested solutions:
- Update the [Profiling docs](https://middy.js.org/docs/best-practices/profiling/) with a paragraph making it clear that requests in parallel can make the measurements invalid/incorrect
- Add an additional parameter to hooks, e.g. `requestId: number | symbol`, so that it's possible to store times/memory during a specific request<sup>1</sup>

**<sup>1</sup>**: A symbol could be a great alternative here, since there's [a proposal where symbols can be used as WeakMap keys](https://github.com/tc39/proposal-symbols-as-weakmap-keys). Using WeakMap would simplify needing to clear up temporary stats stored for each request.

**To Reproduce**
How to reproduce the behaviour:
1. Given a simplified version of the docs example of profiling time for middy, by just measuring handler execution time:
   ```js
   import middy from '@middy/core';
   
   
   let start = 0n;
   
   export const handler = middy(async () => {
     await delay(500);
     return { statusCode: 200 };
   }, {
     beforeHandler: () => start = process.hrtime.bigint(),
     afterHandler: () => console.log(`Duration: ${(process.hrtime.bigint() - start) / 1000000n}ms`)
   });
   
   const callHandler = () => handler({}, {});
   const delay = (ms) => new Promise(res => setTimeout(() => res(undefined), ms));
   
   const res = callHandler();
   await delay(400);
   await Promise.all([res, callHandler()]);
   ```
2. Output will be something like:
   ```
   Duration: 92ms
   Duration: 510ms
   ```

**Expected behaviour**
The measured duration for the handler becomes incorrect. The actual execution time here is close to 500ms, e.g.:
```
Duration: 500ms
Duration: 500ms
```

The profiling example in docs But it does not seem possible to for hooks to measure correct durations when there is no way to uniquely identify requests.

**Environment (please complete the following information):**
 - Node.js: 18
 - Middy: 4.2.6
 - AWS SDK: 3.284.0

**Additional context**

**Comment by csvn** [comment]
I hit upon issues trying to measure execution time for our app. Since there are multiple requests during startup, we got a lot of random results

Was unsure to label this as a documentation bug, or a feature request for hooks to more robustly be able to measure execution time of handlers/middleware for requests without race conditions.

**Comment by willfarrell** [comment]
When lambda receives more than one request at the same time it spins up another sandbox and run's them separately. It's not possible for a single lambda invocation to process more than 1 request at a time that I know of [https://docs.aws.amazon.com/lambda/latest/dg/lambda-concurrency.html]. Can you share how your single lambda is running more than 1 invocation within the same sandbox at the same time.

The timer/memory profilers were developed to test middlewares within this mono-repo. This is why they are not published as packages. I felt others might find them useful as examples or a starting point for writing their own use cases. PR to improve the docs are always welcome.

As you pointed out all most plugin hooks don't pass in the request. This is by design to prevent "meta-middlewares" (using beforeHandler/afterHandler for a middleware) and prevent tampering of the request making debug a nightmare. The `requestEnd` hook does get `request` passed into it to allow enriching of any data collected before sending if off. If you need `request` in the other hook, please open a new issue with your use case for discussion.

**Comment by csvn** [comment]
Thanks for your response!

> When lambda receives more than one request at the same time it spins up another sandbox and run's them separately.
Sorry, I guess I was mistaken here. I thought it scaled up new sandboxes not per request, but based on some amount of load or concurrent requests. Since Node.js is event driven, and a lambda may often be idle while waiting for DB responses etc. Using e.g. express and a "standard" Node server works more like I assumed, so I guess I just had the wrong mental model!

> Can you share how your single lambda is running more than 1 invocation within the same sandbox at the same time.
Our local setup is a bit different from how we deploy it to AWS with AWS-CDK, so I only reproduced it locally but not on Lambda. I just had the wrong assumption here.

> This is why they are not published as packages. I felt others might find them useful as examples or a starting point for writing their own use cases.
Yeah, they were definitely helpful examples, since we saw increased latency when adding a lambda in-front of our app.

>This is by design to prevent "meta-middlewares" (using beforeHandler/afterHandler for a middleware)
Yeah, I understood not passing `request` directly was by design to keep hooks simple. That's why I thought we just needed something to be able to distinguish between multiple requests. But if there is ever only on request at a time then this is not an issue.

Your explanation about no concurrent requests for each lambda might explain why we're seeing longer latency though. We're using [jwks-rsa](https://www.npmjs.com/package/jwks-rsa), and for newly created sandboxes there is no cached public keys and thus requests with their JWT token take longer to verify for concurrent requests (spinning up new instances with empty cache). There are probably underlying reasons to not have concurrent lambda requests, but feels a bit inefficient 😅

It could be helpful with a link to the AWS docs about the lambda concurrency on either the profiling examples page or the hooks page (or maybe I'm just a special case 😇). Anyway, thanks a lot for your quick response!
